src/app/clusters/scenes/ExtensionFieldSetsImpl.cpp

```diff
@@ -0,0 +1,148 @@
+
+ExtensionFieldSetsImpl::ExtensionFieldSetsImpl() : ExtensionFieldSets() {}
+
+CHIP_ERROR ExtensionFieldSetsImpl::Serialize(TLV::TLVWriter & writer) const
+{
+    TLV::TLVType container;
+    ReturnErrorOnFailure(writer.StartContainer(TLV::ContextTag(kTagEFSArrayContainer), TLV::kTLVType_Structure, container));
+    ReturnErrorOnFailure(writer.Put(TLV::ContextTag(kTagEFSFieldNum), static_cast<uint8_t>(this->mFieldNum)));
+    if (!this->IsEmpty())
+    {
+        for (uint8_t i = 0; i < this->mFieldNum; i++)
+        {
+            if (!this->mEFS[i].IsEmpty())
+            {
+                ReturnErrorOnFailure(this->mEFS[i].Serialize(writer));
+            }
+        }
+    }
+
+    return writer.EndContainer(container);
+}
+
+CHIP_ERROR ExtensionFieldSetsImpl::Deserialize(TLV::TLVReader & reader)
+{
+    TLV::TLVType container;
+    ReturnErrorOnFailure(reader.Next(TLV::kTLVType_Structure, TLV::ContextTag(kTagEFSArrayContainer)));
+    ReturnErrorOnFailure(reader.EnterContainer(container));
+
+    ReturnErrorOnFailure(reader.Next(TLV::ContextTag(kTagEFSFieldNum)));
+    ReturnErrorOnFailure(reader.Get(this->mFieldNum));
+
+    if (!this->IsEmpty())
+    {
+        for (uint8_t i = 0; i < this->mFieldNum; i++)
+        {
+            ReturnErrorOnFailure(this->mEFS[i].Deserialize(reader));
+        }
+    }
+
+    return reader.ExitContainer(container);
+}
+
+void ExtensionFieldSetsImpl::Clear()
+{
+    if (!this->IsEmpty())
+    {
+        for (uint8_t i = 0; i < kMaxClusterPerScenes; i++)
+        {
+            this->mEFS[i].Clear();
+        }
+    }
+    this->mFieldNum = 0;
+}
+
+/// @brief Inserts a field Set set into the array of extension field Set sets for a scene entry.
+///        If the same ID is present in the EFS array, it will overwrite it.
+/// @param fieldSet field set to be inserted
+/// @return CHIP_NO_ERROR if insertion worked, CHIP_ERROR_NO_MEMORY if the array is already full
+CHIP_ERROR ExtensionFieldSetsImpl::InsertFieldSet(ExtensionFieldsSet & fieldSet)
+{
+    CHIP_ERROR err             = CHIP_ERROR_NO_MEMORY;
+    uint8_t idPosition         = kInvalidPosition;
+    uint8_t firstEmptyPosition = kInvalidPosition;
+
+    VerifyOrReturnError(fieldSet.mID != kInvalidClusterId, CHIP_ERROR_INVALID_ARGUMENT);
+
+    for (uint8_t i = 0; i < kMaxClusterPerScenes; i++)
+    {
+        if (this->mEFS[i].mID == fieldSet.mID)
+        {
+            idPosition = i;
+            break;
+        }
+
+        if (this->mEFS[i].IsEmpty() && firstEmptyPosition == 0xFF)
+        {
+            firstEmptyPosition = i;
+        }
+    }
+
+    // if found, replace at found position, otherwise at insert first free position, otherwise return error
+    if (idPosition < kMaxClusterPerScenes)
+    {
+        this->mEFS[idPosition] = fieldSet;
+        err                    = CHIP_NO_ERROR;
+    }
+    else if (firstEmptyPosition < kMaxClusterPerScenes)
+    {
+        this->mEFS[firstEmptyPosition] = fieldSet;
+        this->mFieldNum++;
```

Comment:
It was assumed that VerifyOrReturnError(fieldSet.mID != kInvalidClusterId, CHIP_ERROR_INVALID_ARGUMENT); would check for empty field sets as an empty field set would have mID = kInvalidClusterId by default. A check on IsEmpty() will be added.

Reply:
Why would an empty fieldset (i.e. one with no data) have an invalid cluster id?  It could just be that no data was encoded for the cluster, no?

Or am I totally mis-understanding how this works?